server/src/main/java/org/opensearch/common/blobstore/EncryptedBlobContainer.java

```diff
@@ -0,0 +1,200 @@
+        return cryptoManager.createDecryptingStream(inputStream);
+    }
+
+    private EncryptedHeaderContentSupplier  getEncryptedHeaderContentSupplier(String blobName) {
+        return (start, end) -> {
+            byte[] buffer;
+            int length = (int) (end - start + 1);
+            try (InputStream inputStream = blobContainer.readBlob(blobName, start, length)) {
+                buffer = new byte[length];
+                inputStream.readNBytes(buffer, (int) start, buffer.length);
+            }
+            return buffer;
+        };
+    }
+
+    @Override
+    public InputStream readBlob(String blobName, long position, long length) throws IOException {
+        Object encryptionMetadata = cryptoManager.loadEncryptionMetadata(getEncryptedHeaderContentSupplier(blobName));
+        DecryptedRangedStreamProvider decryptedStreamProvider = cryptoManager.createDecryptingStreamOfRange(
+                encryptionMetadata,
+                position,
+                position + length - 1
+        );
+        long adjustedPos = decryptedStreamProvider.getAdjustedRange()[0];
+        long adjustedLength = decryptedStreamProvider.getAdjustedRange()[1] - adjustedPos + 1;
+        InputStream encryptedStream = blobContainer.readBlob(blobName, adjustedPos, adjustedLength);
+        return decryptedStreamProvider.getDecryptedStreamProvider().apply(encryptedStream);
+    }
+
+    @Override
+    public long readBlobPreferredLength() {
+        return blobContainer.readBlobPreferredLength();
+    }
+
+    private void executeWrite(InputStream inputStream, long blobSize,
+                              BiConsumer<InputStream, Long> writeConsumer) {
+        Object cryptoContext = cryptoManager.initEncryptionMetadata();
+        InputStreamContainer streamContainer = new InputStreamContainer(inputStream, blobSize, 0);
+        InputStreamContainer encryptedStream = cryptoManager.createEncryptingStream(cryptoContext, streamContainer);
+        long cryptoLength = cryptoManager.estimateEncryptedLengthOfEntireContent(cryptoContext, blobSize);
+        writeConsumer.accept(encryptedStream.getInputStream(), cryptoLength);
+    }
+
+    @Override
+    public void writeBlob(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+        IOException[] ioException = new IOException[1];
+        executeWrite(inputStream, blobSize, (encryptedStream, encryptedLength) -> {
+            try {
+                blobContainer.writeBlob(blobName, encryptedStream, encryptedLength, failIfAlreadyExists);
+            } catch (IOException ex) {
+                ioException[0] = ex;
+            }
+        });
+        if (ioException[0] != null) {
+            throw ioException[0];
+        }
+    }
+
+    @Override
+    public void writeBlobAtomic(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists) throws IOException {
+        IOException[] ioException = new IOException[1];
```

Comment:
Curious how does an array of IOException work in this case?

Reply:
`executeWrite` is just a reusable piece of code b/w `writeBlob` and `writeBlobAtomic`. It isn't asynchronous and therefore, it is safe to look for a change in ioException after invocation of `executeWrite`. Typically, we use AtomicReference but since this piece is synchronous, I went ahead with a single sized array.

Reply:
Can you use a `org.opensearch.common.CheckedBiConsumer<InputStream, Long, IOException>` instead of a plain `BiConsumer` in `executeWrite()` and directly throw the exception without needing to capture it in a single element array?

Reply:
done